VBA Fields validation

akshayasalvi opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

@puneetlath Can you please review this?

### Details
- Added regex for phone validation
- Added maxLength and keyboardType for fields

### Fixed Issues
<!---
Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing.
Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the issue; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/App/issues/5184

### Tests
1. Tested the phone number with brackets and with long digits
2. Tested ssn with more than 4 digits
3. Tested taxIDNumber with more than 9 digits
4. Tested zipCode with more than 5 digits

<!--- 
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image 
--->

### QA Steps
There are two forms where the validation is added.

**Company Step**
1. Got to `bank-account/company-step`
2. Enter invalid values for `zipCode`, `taxIDNumber` and `phoneNumber`
3. Click on `Save & Continue`
4. Validate that fields show the error


**Requestor Step** 
1. Got to `bank-account/personal-information`
2. Enter invalid values for `ssn`
3. Click on `Save & Continue`
4. Validate that fields show the error


<!--- 
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image 
--->

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<!-- Insert screenshots of your changes on the web platform-->
<img width="368" alt="web-ssn-error" src="https://user-images.githubusercontent.com/57435789/134783468-c20110cc-702a-452d-b4cc-d5ec2d7219c4.png">
<img width="377" alt="web-ssn-valid" src="https://user-images.githubusercontent.com/57435789/134783480-f34d6c00-6887-4a45-9208-f5d3d966fc64.png">
<img width="392" alt="web-validation-errors" src="https://user-images.githubusercontent.com/57435789/134783482-5408c9a6-0318-493c-8eaa-c61a2ea936d2.png">
<img width="399" alt="web-valid-values" src="https://user-images.githubusercontent.com/57435789/134783483-fbe13f6b-0bf1-4b16-ba61-71321aa1b1bf.png">

#### Mobile Web
<!-- Insert screenshots of your changes on the web platform (from a mobile browser)-->
<img width="357" alt="mweb-validation-errors" src="https://user-images.githubusercontent.com/57435789/134783517-a2ae5d74-e09b-4fa5-9600-8ee02f4b2a26.png">
<img width="341" alt="mweb-valid-values-001" src="https://user-images.githubusercontent.com/57435789/134783521-b7a2b965-140e-496d-842f-b5bbd7c9dae9.png">
<img width="344" alt="mweb-valid-values-002" src="https://user-images.githubusercontent.com/57435789/134783522-99d63739-3658-4dc5-99b3-0c3e1ba9bb61.png">
<img width="374" alt="mweb-ssn-error" src="https://user-images.githubusercontent.com/57435789/134783523-7ddeb35e-1e36-434f-b28d-216ed76b3b5c.png">
<img width="419" alt="mweb-ssn-valid" src="https://user-images.githubusercontent.com/57435789/134783524-b56954ef-f21f-4aea-990e-3b5fe9c6ce5d.png">

#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->
<img width="414" alt="desktop-validation-errors" src="https://user-images.githubusercontent.com/57435789/134783533-c8639e82-15ab-4c74-b64a-4aa0316ba2b5.png">
<img width="386" alt="desktop-valid-values" src="https://user-images.githubusercontent.com/57435789/134783535-864995cc-f790-4569-a0c4-adb5b2094864.png">
<img width="374" alt="desktop-ssn-error" src="https://user-images.githubusercontent.com/57435789/134783673-73983f82-8a8b-4091-ac46-d5f251953c06.png">
<img width="376" alt="desktop-ssn-valid" src="https://user-images.githubusercontent.com/57435789/134783675-01e1e0d9-2c03-4d04-9d67-461fb488b4f5.png">


#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->
<img width="356" alt="ios-validation-errors" src="https://user-images.githubusercontent.com/57435789/134783544-546ede5d-b894-43ad-b66b-4c10a7012187.png">
<img width="377" alt="ios-valid-values" src="https://user-images.githubusercontent.com/57435789/134783546-de868491-5f32-4ab7-b249-87e8c626ea09.png">
<img width="349" alt="ios-ssn-error" src="https://user-images.githubusercontent.com/57435789/134783549-585c616e-11f9-463a-a181-44bedb178852.png">
<img width="349" alt="ios-ssn-valid" src="https://user-images.githubusercontent.com/57435789/134783550-b5e6847b-58fa-4275-bd37-963955c3c2d0.png">

#### Android
<!-- Insert screenshots of your changes on the Android platform-->
<img width="326" alt="android-validation-errors" src="https://user-images.githubusercontent.com/57435789/134783557-39c16d97-985c-4ec6-bf45-92add585cdb3.png">
<img width="324" alt="android-valid-values" src="https://user-images.githubusercontent.com/57435789/134783559-ca792646-8fd7-45ae-98ce-da4491191738.png">
<img width="322" alt="android-ssn-error" src="https://user-images.githubusercontent.com/57435789/134783563-d43c8f38-b555-4291-b68f-2b7abb157264.png">
<img width="340" alt="android-ssn-value" src="https://user-images.githubusercontent.com/57435789/134783565-2ee68ac9-5ce2-4d91-b87b-cb8051d4d2ac.png">

akshayasalvi committed:
Added validation for phone, ssn and taxt_id_number

akshayasalvi committed:
Added maxlength and keyboard type for zipcode and phone number

akshayasalvi committed:
Added maxlength to zipcode

akshayasalvi commented:
@puneetlath We might have to update the placeholder for `Phone number` field. At the moment it says `10 digits, no hyphens`. Can you or @SofiedeVreese help me with the new placeholder that I can update?

Luke9389 commented:
@akshayasalvi Regarding the phone number placeholder, I think it's fine to just use `Phone Number (xxx)xxx-xxxx` .

Also, I think you should use `Str.isValidPhone` from `expensify-common` in the `validate()` method on this form.

Luke9389 commented:
Hey @akshayasalvi, great job so far on this issue. 👍 

We are wanting to push through all issues related to this flow very quickly, so it's possible that I'll end up taking this PR and finishing it personally. Rest assured, you will still get paid as if you completed this issue. I'll check back in tomorrow and make a determination then.

akshayasalvi commented:
> @akshayasalvi Regarding the phone number placeholder, I think it's fine to just use `Phone Number (xxx)xxx-xxxx` .

Same placeholder in Spanish too?

> Also, I think you should use `Str.isValidPhone` from `expensify-common` in the `validate()` method on this form.

I had seen that but it seems our current `isValidPhone` only allows digits and not parenthesis? I had even raised it in the discussion here https://github.com/Expensify/App/issues/5184#issuecomment-922929708


> Hey @akshayasalvi, great job so far on this issue. 👍

Thank you

> We are wanting to push through all issues related to this flow very quickly, so it's possible that I'll end up taking this PR and finishing it personally. Rest assured, you will still get paid as if you completed this issue. I'll check back in tomorrow and make a determination then.

Okay both the changes that you've mentioned have follow-up questions. Can you help so that I'll update the PR at the earliest?

Luke9389 commented:
> Same placeholder in Spanish too?

That's a good point. For Spanish, lets do `(prefijo) + (número)`.

Luke9389 commented:
> I had seen that but it seems our current isValidPhone only allows digits and not parenthesis? I had even raised it in the discussion here #5184 (comment)

You also offered taking out the parentheses in that comment, which sounds like a fine solution to me. We don't need to make that a separate method, just do it in one line.

Luke9389 commented on src/CONST.js:
```diff
@@ -419,6 +425,7 @@ const CONST = {
         US_PHONE: /^\+1\d{10}$/,
         DIGITS_AND_PLUS: /^\+?[0-9]*$/,
         PHONE_E164_PLUS: /^\+?[1-9]\d{1,14}$/,
+        PHONE_WITH_SPECIAL_CHARS: /^[+]*[(]{0,1}[0-9]{1,3}[)]{0,1}[-\s\\./0-9]*$/,
```
Didn't see this originally, you can ignore my suggestion to use `Str.isValidPhone`

However, this doesn't limit the number of digits after the area code. I'm not sure if it's best to always rely on using `MAX_LENGTH` in combination with `PHONE_WITH_SPECIAL_CHARS`. Does that make sense?

akshayasalvi replied:
What is max length of the digits after the area code?

Luke9389 commented on src/libs/ValidationUtils.js:
```diff
@@ -69,6 +69,15 @@ function isValidAge(date) {
+ * @returns {Boolean}
+ */
+function isValidPhoneWithSpecialChars(phoneNumber) {
+    return CONST.REGEX.PHONE_WITH_SPECIAL_CHARS.test(phoneNumber);
```
This doesn't test for max length. I know we do that on the input, but  I worry that this will be used without that.

Luke9389 reviewed (commented):
I think this is looking really good. Let me know what you think about my concern with de-coupling the max length of phone numbers and their symbols.

akshayasalvi committed:
Added translations for phone number placeholder

akshayasalvi committed:
Updated regex validation for length and added length check in function

akshayasalvi committed:
Merge branch 'main' of github-personal:akshayasalvi/App into vba-flow-validation

# Conflicts:
#	src/pages/ReimbursementAccount/CompanyStep.js
#	src/pages/ReimbursementAccount/IdentityForm.js

akshayasalvi commented:
> I think this is looking really good. Let me know what you think about my concern with de-coupling the max length of phone numbers and their symbols.

@Luke9389 I've updated the regex with length check on the numbers. I've also added an additional max length check in the function itself.

Also, I saw two conflicts where I've picked my changes. I insist on keeping my change `keyboardType={CONST.KEYBOARD_TYPE.NUMERIC}` whereas the incoming change had it set as `KEYBOARD_TYPE.PHONE_PAD` with fields like `zipCode` and `ssnLast4` being numbers only.

Luke9389 commented on src/CONST.js:
```diff
@@ -420,12 +419,12 @@ const CONST = {
         DIGITS_AND_PLUS: /^\+?[0-9]*$/,
         PHONE_E164_PLUS: /^\+?[1-9]\d{1,14}$/,
-        PHONE_WITH_SPECIAL_CHARS: /^[+]*[(]{0,1}[0-9]{1,3}[)]{0,1}[-\s\\./0-9]*$/,
+        PHONE_WITH_SPECIAL_CHARS: /^[+]*[(]{0,1}[0-9]{1,3}[)]{0,1}[-\s\\./0-9]{0,11}$/,
```
I think this brings us to a max of 14 digits total, right? I think we want to bump this to `{0,12}`

akshayasalvi replied:
Sorry about this. Updated.

Luke9389 commented on src/CONST.js:
```diff
@@ -77,7 +77,6 @@ const CONST = {
             TAX_ID_NUMBER: 9,
             SSN: 4,
             ZIP_CODE: 5,
-            PHONE_NUMBER: 15,
```
I don't understand why this was moved out of `BANK_ACCOUNT.MAX_LENGTH`. Would you explain that for me?

akshayasalvi replied:
I thought phone number being a generic field we might use it in more than one modules, so I've moved it to root object.

Luke9389 replied:
Ok, I think that's reasonable 👍

Luke9389 reviewed (commented):
Getting closer, just a few more questions. Thanks!

akshayasalvi committed:
Updated length to 12

akshayasalvi committed:
Merge branch 'main' of github-personal:akshayasalvi/App into vba-flow-validation

# Conflicts:
#	src/libs/ValidationUtils.js
#	src/pages/ReimbursementAccount/CompanyStep.js

akshayasalvi commented:
PR updated

Luke9389 reviewed (dismissed):
Before merging I'd like to verify that you've tested these changes since the most recent commit. I ask all my contributors this question because it's easy to assume our commits don't break anything, and then they cause regressions. Lets get you paid on time :)

akshayasalvi commented:
Yes tested :)

akshayasalvi commented:
@Luke9389 @MonilBhavsar Anything that needs to be checked from my side here?

Luke9389 commented:
The only thing @akshayasalvi is that there's a merge conflict in this PR right now. Can you resolve that pls?

akshayasalvi committed:
Merge branch 'main' of github-personal:akshayasalvi/App into vba-flow-validation

# Conflicts:
#	src/libs/ValidationUtils.js

akshayasalvi commented:
@Luke9389 Resolved and tested again

Luke9389 commented:
@akshayasalvi Looks like there's another conflict.

akshayasalvi commented:
> @akshayasalvi Looks like there's another conflict.

Ohh no :(

Luke9389 commented:
Heh, it's OK! No worries, just be sure you're merging with the most recent version of our `main` branch. 
Usually I run `git fetch` to help with this.

akshayasalvi committed:
Merge branch 'main' of github-personal:akshayasalvi/App into vba-flow-validation

# Conflicts:
#	src/pages/ReimbursementAccount/IdentityForm.js

akshayasalvi commented:
Resolved again and tested

> Heh, it's OK! No worries, just be sure you're merging with the most recent version of our `main` branch. Usually, I run `git fetch` to help with this. (Though, if you want to be 200% sure, checkout main and run `git fetch && git pull`)

Yeah, I do it every time. It is just other changes are getting merged and resulting in conflicts. I am guessing these screens are under active development.

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1312890262) to staging by @Luke9389 in version: 1.1.5-2 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

puneetlath commented:
Hm, that seems like a separate issue rather than a regression caused by this PR. This PR just enforces that what is entered is the right type and length. But doesn't do any validation on whether the numbers entered are valid. So I don't think we need to block this PR based on that.

marcaaron commented:
Just noticed the regex initially added here allows people to enter phone numbers that are only 3 digits ?

Seems like something is wrong? Or is that expected..?

![2021-10-15_07-25-11](https://user-images.githubusercontent.com/32969087/137528425-75dd8c26-41df-4a68-a215-396988eac6d0.png)

marcaaron commented:
Yes, that's good. But what 3 digit number are they entering that we should accept as a phone number?
